junit-vintage-engine/src/test/java/org/junit/vintage/engine/DynamicTestsHelperRun.java

```diff
@@ -0,0 +1,105 @@
+ *
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+
+package org.junit.vintage.engine;
+
+import java.util.HashMap;
+import java.util.Map;
+
+import org.junit.Assert;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.junit.platform.engine.TestExecutionResult;
+import org.junit.platform.engine.discovery.DiscoverySelectors;
+import org.junit.platform.launcher.Launcher;
+import org.junit.platform.launcher.LauncherDiscoveryRequest;
+import org.junit.platform.launcher.TestExecutionListener;
+import org.junit.platform.launcher.TestIdentifier;
+import org.junit.platform.launcher.core.LauncherDiscoveryRequestBuilder;
+import org.junit.platform.launcher.core.LauncherFactory;
+
+class DynamicTestsHelperRun {
+
+	private Launcher launcher;
+	private Map<State, Integer> callCounts;
+	private LauncherDiscoveryRequestBuilder requestBuilder;
+	enum State {
+		failed, success,
+	}
+
+	@BeforeEach
+	void setUpLauncher() {
+		this.launcher = LauncherFactory.create();
+		this.callCounts = new HashMap<>();
+		this.requestBuilder = LauncherDiscoveryRequestBuilder.request();
+		launcher.registerTestExecutionListeners(new TestListenerMock(callCounts));
+	}
+
+	@Test
+	void succeeding() {
+		requestBuilder.selectors(DiscoverySelectors.selectClass(JUnit4ParametrizedExample.class));
+		LauncherDiscoveryRequest discoveryRequest = requestBuilder.build();
+
+		launcher.discover(discoveryRequest);
+		launcher.execute(discoveryRequest);
+
+		final HashMap<State, Integer> expectedCallCounts = new HashMap<>();
+		expectedCallCounts.put(State.success, 3);
+		expectedCallCounts.put(State.failed, 9);
+
+		Assert.assertEquals(expectedCallCounts, callCounts);
+	}
+
+	@Test
+	void failing() {
+		requestBuilder.selectors(DiscoverySelectors.selectMethod(JUnit4ParametrizedExample.class, "test1"));
+		LauncherDiscoveryRequest discoveryRequest = requestBuilder.build();
+
+		launcher.discover(discoveryRequest);
+		launcher.execute(discoveryRequest);
+
+		final HashMap<State, Integer> expectedCallCounts = new HashMap<>();
+		expectedCallCounts.put(State.failed, 3);
+
+		Assert.assertEquals(expectedCallCounts, callCounts);
+	}
+
+	private static State calcState(TestExecutionResult result) {
+		TestExecutionResult.Status status = result.getStatus();
+		switch (status) {
+			case FAILED: {
+				return State.failed;
+			}
+			case SUCCESSFUL: {
+				return State.success;
+			}
+			default: {
+				throw new IllegalStateException("unknown state");
+			}
+		}
+	}
+
+	private class TestListenerMock implements TestExecutionListener {
+		private final Map<State, Integer> callCounts;
+
+		TestListenerMock(Map<State, Integer> callCounts) {
+			this.callCounts = callCounts;
+		}
+
+		@Override
+		public void executionFinished(TestIdentifier identifier, TestExecutionResult result) {
+			if (identifier.isTest()) {
+				State state = calcState(result);
+				callCounts.putIfAbsent(state, 0);
+				callCounts.put(state, callCounts.get(state) + 1);
```

Comment:
Can't you combine the previous two lines into a single call to `java.util.Map.merge(K, V, BiFunction<? super V, ? super V, ? extends V>)`?